File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    answer = 0
    
    start = min(diffs)
    end = max(diffs)
    length = len(diffs) # 길이
    
    while start <= end:
        mid = (start + end) // 2
        total = 0
        
        for i in range(length):
            if i == 0:
                prev = 0 # 현재가 첫번째라면 prev 0
            else:
                prev = times[i-1] 
            
            cur = times[i]
            
            if mid >= diffs[i]:
                total += cur
            else:
                total += cur + ((diffs[i] - mid) * (prev + cur)) 
        
        if total > limit: # total > limit이면 level을 더 올려야함
            start = mid + 1
        else: # total <= limit 이면 level을 더 낮춰야함 
            end = mid - 1
        
    return start

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
import java.util.*;

class Solution {
    public int solution(int[] diffs, int[] times, long limit) {
        int start = Arrays.stream(diffs).min().getAsInt();
        int end = Arrays.stream(diffs).max().getAsInt();
        int size = times.length;
        
        while (start <= end) {
            int mid = (start + end) / 2;
            long total = 0;
            
            for (int i = 0; i < size; i++) {
                int cur = times[i];
                int prev;
                
                if (i == 0) {
                    prev = 0;
                } else {
                    prev = times[i-1];
                }
                
                if (mid >= diffs[i]) {
                    total += cur;
                } else {
                    total += cur + ((diffs[i] - mid) * (cur + prev));
                }
            }
            
            if (total > limit) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
            
        }
        
        return end + 1;
    }
}